server/src/main/java/org/jboss/as/server/ServerEnvironment.java

```diff
@@ -1242,4 +1277,16 @@ private File[] getFilesFromProperty(final String name, final Properties props) {
     ManagedAuditLogger createAuditLogger() {
         return new ManagedAuditLoggerImpl(getProductConfig().resolveVersion(), true);
     }
+
+    ServerEnvironment recalculateForReload(RunningModeControl runningModeControl) {
+        if (runningModeControl.isReloaded()) {
+            Stability stability = runningModeControl.getReloadedStability() != null ? runningModeControl.getReloadedStability() : this.stability;
+            if (stability != this.stability) {
+                System.setProperty(ProcessEnvironment.STABILITY, stability.toString());
+
+                return new ServerEnvironment(this, stability);
```

Comment:
I have some concerns about creating new ServerEnvironment objects here, which inturns forces us to create a new Configuration object as well.

I wonder if it would be better to use the current reference, and instead of creating a new ServerEnvironment object, updating the new stability level on the current ServerEnvironment reference. 

That could be done on the ServerProcessReloadHandler#doReload, by creating a new method on ServerEnvironment to update it, the current Stability class field is final, but we can remove this restriction. These handlers should have the single ServerEnvironment instance created when the process starts and stored in the Bootstrap.Configuration class.

That might remove the need to create these new objects. I think there are no problems with the current approach, but as soon as the original Bootstrap.Configuration or ServerEnvironment is shared across other instances outside of the root service that is reloaded, it could lead into issues having two different objects around.

Reply:
+1 @yersan. Good point.

If we add mutator method it will need to be public, so please @Deprecate it and file a Blocker WFCORE to move the handlers that deal with these objects out of the 'operations' package and into the root package, where they can access package protected mutators.

The 'operations' package has been a minor pain point for a long time because it forces public visibility on things just for the sake of arbitrary categorization of classes. We can use this as an excuse to tidy this up a bit.